File: dev/pytools/registry.py
"""The stand registry (dev/stand/registry.yaml): every name the tests and the gate use."""

from __future__ import annotations

import functools
import os
import re
from pathlib import Path

import yaml

PATH = Path(__file__).resolve().parents[1] / "stand" / "registry.yaml"


@functools.lru_cache(maxsize=1)
def load() -> dict:
    """The parsed registry."""
    return yaml.safe_load(PATH.read_text())


def source(name: str) -> dict:
    """`{container, url}` of one stand source (`postgres`, `mongo_rs`, …)."""
    return load()["sources"][name]


def bq_tmp(name: str) -> str:
    """A disposable BigQuery dataset name — the only kind the sweep may drop."""
    return load()["bigquery"]["tmp_prefix"] + name


def orphaned(name: str) -> bool:
    """Is `name` a test object whose creating process is gone?"""
    m = re.search(load()["test_object_suffix"], name)
    if not m:
        return False
    try:
        os.kill(int(m.group(1)), 0)
    except ProcessLookupError:
        return True
    except (PermissionError, OverflowError, ValueError):
        return False
    return False


def _self_test() -> None:
    assert source("postgres")["container"] == "rivet-postgres-1"
    assert bq_tmp("gate").startswith("rivet_tmp_")
    assert not orphaned("users"), "a persistent fixture is never an orphan"
    assert not orphaned(f"t_{os.getpid()}_3"), "a live process's object is spared"
    dead = 2**22 + 12345
    assert orphaned(f"t_{dead}_1"), "a dead pid's object is swept"
    print("registry self-test ok")


if __name__ == "__main__":
    _self_test()

File: tests/common/registry.rs
//! The stand registry (`dev/stand/registry.yaml`): every name the tests and the gate use.
#![allow(dead_code)]

use std::sync::OnceLock;

const REGISTRY_YAML: &str = include_str!("../../dev/stand/registry.yaml");

/// The parsed registry, compiled into the test binary.
pub fn stand() -> &'static serde_yaml_ng::Value {
    static R: OnceLock<serde_yaml_ng::Value> = OnceLock::new();
    R.get_or_init(|| {
        serde_yaml_ng::from_str(REGISTRY_YAML).expect("dev/stand/registry.yaml parses")
    })
}

fn text(path: &[&str]) -> &'static str {
    path.iter()
        .fold(stand(), |v, k| &v[*k])
        .as_str()
        .unwrap_or_else(|| panic!("dev/stand/registry.yaml has no string at {path:?}"))
}

/// Container name of a stand source (`postgres`, `mongo_rs`, …).
pub fn stand_container(source: &str) -> &'static str {
    text(&["sources", source, "container"])
}

/// Connection URL of a stand source.
pub fn stand_url(source: &str) -> &'static str {
    text(&["sources", source, "url"])
}

/// `host:port` of a stand source, parsed from its URL.
pub fn stand_host_port(source: &str) -> &'static str {
    let url = stand_url(source);
    let rest = url.split("://").nth(1).expect("registry url has a scheme");
    let rest = rest.rsplit('@').next().unwrap_or(rest);
    let end = rest.find(['/', '?']).unwrap_or(rest.len());
    &rest[..end]
}

/// The one permanent BigQuery dataset (live tests and the harness).
pub fn stand_bq_e2e() -> &'static str {
    text(&["bigquery", "e2e"])
}

/// The BigQuery location every test dataset is created in.
pub fn stand_bq_location() -> &'static str {
    text(&["bigquery", "location"])
}

/// A disposable BigQuery dataset name — the only kind `make sweep-test-cloud` drops.
pub fn stand_bq_tmp(name: &str) -> String {
    format!("{}{name}", text(&["bigquery", "tmp_prefix"]))
}

/// The seeded 1M-row fixture database the harness reads.
pub fn stand_bench_db() -> &'static str {
    stand()["databases"]
        .as_sequence()
        .and_then(|s| {
            s.iter()
                .filter_map(|v| v.as_str())
                .find(|d| d.ends_with("_bench"))
        })
        .expect("registry lists a *_bench database")
}
